src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -0,0 +1,309 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading.Channels;
+using Aspire.Dashboard.Utils;
+using Aspire.V1;
+using Grpc.Core;
+using Grpc.Net.Client;
+
+namespace Aspire.Dashboard.Model;
+
+/// <summary>
+/// Implements gRPC client that communicates with a resource server, populating data for the dashboard.
+/// </summary>
+/// <remarks>
+/// An instance of this type is created per service call, so this class should not hold onto any state
+/// expected to live longer than a single RPC request. In the case of streaming requests, the instance
+/// lives until the stream is closed.
+/// </remarks>
+internal sealed class DashboardClient : IDashboardClient
+{
+    private const string DashboardServiceUrlVariableName = "DOTNET_DASHBOARD_GRPC_ENDPOINT_URL";
+    private const string DashboardServiceUrlDefaultValue = "http://localhost:18999";
+
+    private readonly Dictionary<string, ResourceViewModel> _resourceByName = new(StringComparers.ResourceName);
+    private readonly CancellationTokenSource _cts = new();
+    private readonly object _lock = new();
+
+    private ImmutableHashSet<Channel<ResourceViewModelChange>> _outgoingChannels = [];
+    private string? _applicationName;
+    private string? _applicationVersion;
+
+    private const int StateNone = 0;
+    private const int StateInitialized = 1;
+    private const int StateDisposed = 2;
+    private int _state;
+
+    private TaskCompletionSource<DashboardService.DashboardServiceClient> _client = new();
+
+    private void EnsureInitialized()
+    {
+        var priorState = Interlocked.CompareExchange(ref _state, StateInitialized, comparand: StateNone);
+
+        if (priorState is not StateNone)
+        {
+            ObjectDisposedException.ThrowIf(priorState is StateDisposed, this);
+            return;
+        }
+
+        var address = GetAddressUri(DashboardServiceUrlVariableName, DashboardServiceUrlDefaultValue);
+
+        _ = Task.Run(() => ConnectAndStayConnectedAsync(address, _cts.Token), _cts.Token);
+
+        async Task ConnectAndStayConnectedAsync(Uri address, CancellationToken cancellationToken)
```

Comment:
What you're doing here with recreating the channel and client isn't necessary. A channel will automatically connect and reconnect when required.

Retrying should focus on the call to the server. It is the what needs to be recreated on error.

Reply:
Oh that's good to know. I'll take another look. There's more work I want to do around reconnects, so I might have a follow up for that.